File: Searching_and_Sorting1/MedianInRow-wiseSortedMatrix.cpp
// https://www.geeksforgeeks.org/problems/median-in-a-row-wise-sorted-matrix1527/1


#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int countLessEqual(vector<int>& row, int target) {
    return upper_bound(row.begin(), row.end(), target) - row.begin();
}

int findMedian(vector<vector<int>>& matrix) {
    int n = matrix.size(), m = matrix[0].size();
    int low = 1, high = 2000;

    while (low < high) {
        int mid = (low + high) / 2;
        int count = 0;

        for (int i = 0; i < n; i++) {
            count += countLessEqual(matrix[i], mid);
        }

        if (count <= (n * m) / 2)
            low = mid + 1;
        else
            high = mid;
    } 
    return low;
}

File: Searching_and_Sorting1/MedianInRow-wiseSortedMatrix.js
// https://www.geeksforgeeks.org/problems/median-in-a-row-wise-sorted-matrix1527/1

function findMedian(matrix) {
    const n = matrix.length;
    const m = matrix[0].length;

    let low = 1, high = 2000;

    const countLessEqual = (row, target) => {
        let l = 0, r = row.length;
        while (l < r) {
            let mid = Math.floor((l + r) / 2);
            if (row[mid] <= target) l = mid + 1;
            else r = mid;
        }
        return l;
    };

    while (low < high) {
        let mid = Math.floor((low + high) / 2);
        let count = 0;
        for (let row of matrix) {
            count += countLessEqual(row, mid);
        }

        if (count <= Math.floor((n * m) / 2)) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }

    return low;
}
